northstar/src/runtime/token.rs

```diff
@@ -4,197 +4,223 @@ use hmac::{
 }
 
-pub type Timestamp = u128;
+type HmacSha256 = hmac::Hmac<Sha256>;
+type Hmac = CtOutput<HmacSha256>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum VerificationResult {
+    /// Verification succeeded
+    Ok,
+    /// Verification failed
+    Invalid,
+    /// Token is expired
+    Expired,
+    /// Token time is in the future
+    Future,
+}
 
-#[derive(Debug, PartialEq, Serialize, Deserialize)]
-pub struct Token {
-    time: Timestamp,
+/// Token instance
+#[derive(Clone, PartialEq)]
+pub(crate) struct Token {
+    time: time::Duration,
     hmac: Hmac,
 }
 
-type HmacSha256 = hmac::Hmac<Sha256>;
-
 impl Token {
-    pub fn new(usage: &str) -> Token {
-        let time = Token::current_time();
-        let hmac = Token::calculate_hmac(&time, usage);
-        Token { time, hmac }
-    }
-
-    pub fn verify(&self, usage: &str) -> bool {
-        let hmac = Token::calculate_hmac(&self.time, usage);
-        let is_authentic = self.hmac == hmac; // constant time comparison ensured by CtOutput
-        let is_expired = Token::current_time()
-            .checked_sub(self.time)
-            .map_or(true, |age| age > TOKEN_EXPIRED_THRESHOLD);
-        is_authentic && !is_expired
-    }
-
-    pub fn write<T: std::io::Write>(&self, output: &mut T) -> std::io::Result<()> {
-        let time = self.time.to_be_bytes();
-        let hmac = self.hmac.clone().into_bytes();
-        output.write_all(&time)?;
-        output.write_all(&hmac)?;
-        Ok(())
+    /// Create a new token
+    pub fn new<U, T, S>(user: U, target: T, shared: S) -> Token
+    where
+        U: AsRef<[u8]>,
+        T: AsRef<[u8]>,
+        S: AsRef<[u8]>,
+    {
+        let now = now();
+        let hmac = calculate_hmac(&now, user.as_ref(), target.as_ref(), shared.as_ref());
+        Token { time: now, hmac }
+    }
+
+    /// Verify that `shared` matches the token
+    pub fn verify<U, T, S>(&self, user: U, target: T, shared: S) -> VerificationResult
+    where
+        U: AsRef<[u8]>,
+        T: AsRef<[u8]>,
+        S: AsRef<[u8]>,
+    {
+        let now = now();
+
+        if now < self.time {
+            VerificationResult::Future
+        } else if now - self.time > TOKEN_EXPIRED_THRESHOLD {
+            VerificationResult::Expired
+        } else if calculate_hmac(&self.time, user.as_ref(), target.as_ref(), shared.as_ref())
+            == self.hmac
+        {
+            VerificationResult::Ok
+        } else {
+            VerificationResult::Invalid
+        }
     }
+}
 
-    fn calculate_hmac(time: &Timestamp, usage: &str) -> Hmac {
-        let mut hasher = HmacSha256::new_from_slice(MAC_KEY.as_slice())
-            .expect("Failed to create SHA-256 hasher");
-        hasher.update(&time.to_be_bytes());
-        hasher.update(usage.as_bytes());
-        Hmac(hasher.finalize())
-    }
+#[cfg(not(test))]
+fn now() -> time::Duration {
+    time::Duration::from_secs(time::Instant::now().elapsed().as_secs())
```

Comment:
I think `time::Instant::now().elapsed()` always returns zero meaning the whole function always returns a duration of zero.

Reply:
Aeh. Right. Will fix that!